#' assemble_tracks
#'
#' @param plot_list List. Contains ggplots and/or grobs. These should have have been
#'   generated by ssvTracks::track_* family of functions or be appropriate for
#'   visualization in the same genomic region defined by query_gr.
#' @param query_gr GRanges. Most often the same query_gr used to generate tracks.  A smaller region to view in detail is also valid.
#' @param rel_heights Passed to rel_heights in cowplot::plot_grid. Controls relative heights of component plots.  Should be same length as plot_list.
#'
#' @return A combined ggplot of tracks stacked vertically with consistent x-axes.
#' @export
#'
#' @examples
#' gtf_file = system.file(package = "ssvTracks", "extdata/ESR1_exons.gtf", mustWork = TRUE)
#' ref_gr = rtracklayer::import.gff(gtf_file)
#' query_gr = range(ref_gr)
#'
#' p_ref = track_gene_transcripts(ref_gr,
#'   sel_gene_name = "ESR1",
#'   intron_thickness = 2,
#'   exon_height = .8,
#'   show_tss = TRUE)
#'
#' bam_files = dir(system.file(package = "ssvTracks", "extdata"), full.names = TRUE, pattern = "^M.+RNA.+bam$")
#' bam_files = bam_files[!grepl("_wt_", bam_files) & !grepl("_RUNX1-ko1_", bam_files)]
#' bam_files_df = data.frame(
#'      file = bam_files,
#'      sample = sub("_.+", "", basename(bam_files)))
#' p_rna = track_rna.SE(bam_files_df, query_gr = query_gr, fill_mapping = "black")
#'
#' assemble_tracks(list(p_rna, p_ref), query_gr)
#'
#'
#'
assemble_tracks = function(plot_list, query_gr, rel_heights = rep(1, length(plot_list))){
  stopifnot(length(query_gr) == 1)
  suppressMessages({
    # set x-limits on all plots
    xlim = c(start(query_gr), end(query_gr))
    if(as.character(strand(query_gr)) == "-") xlim = rev(xlim)
    for(i in seq(1, length(plot_list))){
      plot_list[[i]] = plot_list[[i]] +
        coord_cartesian(xlim = xlim)
    }
    # remove x-axis labels from all plots but final
    if(length(plot_list) > 1){
      for(i in seq(1, length(plot_list) - 1)){
        plot_list[[i]] = plot_list[[i]] +
          labs(x = "") +
          theme(axis.text.x = element_blank())
      }
    }
    # add axis labels to final plot
    plot_list[[length(plot_list)]] = plot_list[[length(plot_list)]] +
      labs(x = paste(as.character(seqnames(query_gr)), "kbp")) +
      scale_x_continuous(labels = function(x)x/1e3)
    plot_list = sync_width(plot_list)
    pg = cowplot::plot_grid(plotlist = plot_list, ncol = 1,
                            rel_heights = rel_heights, scale = 1)
  })
  pg
}



#' sync_width
#'
#' @param grob_list list containing ggplot and/or grobs
#'
#' @return list of grobs with hortizontal margins set to equal sizes to allow stacking.
#' @export
#'
#' @examples
#' gtf_file = system.file(package = "ssvTracks", "extdata/ESR1_exons.gtf", mustWork = TRUE)
#' ref_gr = rtracklayer::import.gff(gtf_file)
#' query_gr = range(ref_gr)
#'
#' p_ref = track_gene_transcripts(ref_gr,
#'   sel_gene_name = "ESR1",
#'   intron_thickness = 2,
#'   exon_height = .8,
#'   show_tss = TRUE)
#'
#' bam_files = dir(system.file(package = "ssvTracks", "extdata"), full.names = TRUE, pattern = "^M.+RNA.+bam$")
#' bam_files = bam_files[!grepl("_wt_", bam_files) & !grepl("_RUNX1-ko1_", bam_files)]
#' bam_files_df = data.frame(
#'      file = bam_files,
#'      sample = sub("_.+", "", basename(bam_files)))
#' p_rna = track_rna.SE(bam_files_df, query_gr = query_gr, fill_mapping = "black")
#'
#' glist = sync_width(list(p_rna, p_ref))
#' #this is basically what assemble_tracks does
#' cowplot::plot_grid(plotlist = glist, ncol = 1)
sync_width = function(grob_list){
  stopifnot(class(grob_list) == "list")
  is_ok = sapply(grob_list, function(x){
    "ggplot" %in% class(x) | "grob" %in% class(x)
  })
  stopifnot(all(is_ok))
  my_grobs = lapply(grob_list, function(x){
    if(grid::is.grob(x)){
      x
    }else{
      ggplotGrob(x)
    }
  })

  my_widths = lapply(my_grobs, function(gt){
    gt$widths
  })
  maxWidth = my_widths[[1]]
  if(length(my_widths) > 1){
    for(i in 2:length(my_widths)){
      maxWidth = grid::unit.pmax(maxWidth, my_widths[[i]])
    }
  }
  for(j in 1:length(my_grobs)){
    my_grobs[[j]]$widths = maxWidth
  }
  my_grobs
}

#' sync_height
#'
#' @param grob_list list containing ggplot and/or grobs
#'
#' @return list of grobs with hortizontal margins set to equal sizes to allow stacking.
#' @export
#'
#' @examples
#' gtf_file = system.file(package = "ssvTracks", "extdata/ESR1_exons.gtf", mustWork = TRUE)
#' ref_gr = rtracklayer::import.gff(gtf_file)
#' query_gr = range(ref_gr)
#'
#' p_ref = track_gene_transcripts(ref_gr,
#'   sel_gene_name = "ESR1",
#'   intron_thickness = 2,
#'   exon_height = .8,
#'   show_tss = TRUE)
#'
#' bam_files = dir(system.file(package = "ssvTracks", "extdata"), full.names = TRUE, pattern = "^M.+RNA.+bam$")
#' bam_files = bam_files[!grepl("_wt_", bam_files) & !grepl("_RUNX1-ko1_", bam_files)]
#' bam_files_df = data.frame(
#'      file = bam_files,
#'      sample = sub("_.+", "", basename(bam_files)))
#' p_rna = track_rna.SE(bam_files_df, query_gr = query_gr, fill_mapping = "black")
#'
#' query_gr_small = resize(query_gr, 5e4)
#' xlim_small = c(start(query_gr_small), end(query_gr_small))
#' glist = sync_height(list(
#'   p_rna,
#'   p_rna +
#'     coord_cartesian(xlim = xlim_small) +
#'     labs(title = "zoom in") +
#'     theme(plot.title = element_text(size = 60))
#' ))
#'
#' #this isn't a great example, but notice the plot on the left leaves space for the large title on the right.
#' cowplot::plot_grid(plotlist = glist, nrow = 1)
sync_height = function(grob_list){
  stopifnot(class(grob_list) == "list")
  is_ok = sapply(grob_list, function(x){
    "ggplot" %in% class(x) | "grob" %in% class(x)
  })
  stopifnot(all(is_ok))
  my_grobs = lapply(grob_list, function(x){
    if(grid::is.grob(x)){
      x
    }else{
      ggplotGrob(x)
    }
  })

  my_widths = lapply(my_grobs, function(gt){
    gt$heights
  })
  maxWidth = my_widths[[1]]
  if(length(my_widths) > 1){
    for(i in 2:length(my_widths)){
      maxWidth = grid::unit.pmax(maxWidth, my_widths[[i]])
    }
  }
  for(j in 1:length(my_grobs)){
    my_grobs[[j]]$heights = maxWidth
  }
  my_grobs
}
